shiny/ui/_chat.py

```diff
@@ -0,0 +1,782 @@
+        return chat_ui(
+            id=self.id,
+            placeholder=placeholder,
+            width=width,
+            height=height,
+            fill=fill,
+            **kwargs,
+        )
+
+    @overload
+    def on_user_submit(
+        self, fn: SubmitFunction | SubmitFunctionAsync
+    ) -> reactive.Effect_: ...
+
+    @overload
+    def on_user_submit(
+        self,
+    ) -> Callable[[SubmitFunction | SubmitFunctionAsync], reactive.Effect_]: ...
+
+    def on_user_submit(
+        self, fn: SubmitFunction | SubmitFunctionAsync | None = None
+    ) -> (
+        reactive.Effect_
+        | Callable[[SubmitFunction | SubmitFunctionAsync], reactive.Effect_]
+    ):
+        """
+        Define a function to invoke when user input is submitted.
+
+        Apply this method as a decorator to a function (`fn`) that should be invoked when the
+        user submits a message. The function should take no arguments.
+
+        In many cases, the implementation of `fn` should do at least the following:
+            1. Call `.get_messages()` to obtain the current chat history.
+            2. Generate a response based on those messages.
+            3. Append the response to the chat history using `.append_message()` or
+              `.append_message_stream()`.
+
+        Parameters
+        ----------
+        fn
+            A function to invoke when user input is submitted.
+
+        Note
+        ----
+        This method creates a reactive effect that only gets invalidated when the user
+        submits a message. Thus, the function `fn` can read other reactive dependencies,
+        but it will only be re-invoked when the user submits a message.
+        """
+
+        def create_effect(fn: SubmitFunction | SubmitFunctionAsync):
+            afunc = _utils.wrap_async(fn)
+
+            @reactive.effect
+            @reactive.event(self.get_user_input)
+            async def handle_user_input():
+                if self._suspend_input_handler:
+                    from .. import req
+
+                    req(False)
+                try:
+                    await afunc()
+                except Exception as e:
+                    await self._raise_exception(e)
+
+            self._effects.append(handle_user_input)
+
+            return handle_user_input
+
+        if fn is None:
+            return create_effect
+        else:
+            return create_effect(fn)
+
+    async def _raise_exception(
+        self,
+        e: BaseException,
+    ) -> None:
+        if self.on_error == "unhandled":
+            raise e
+        else:
+            await self._remove_loading_message()
+            sanitize = self.on_error == "sanitize"
+            raise NotifyException(str(e), sanitize=sanitize)
+
+    def get_messages(
+        self,
+        *,
+        token_limits: tuple[int, int] | None = (4096, 1000),
+        apply_user_transform: bool = True,
+        apply_assistant_transform: bool = False,
+    ) -> tuple[ChatMessage, ...]:
```

Comment:
The return type of this function is `tuple[ChatMessage, ...]`. Looking at the type annotations for anthropic and openai and ollama, it looks to me like the message formats between them are similar, but they diverge in how they represent content other than basic text (like images, function calls, and other metadata).

I think we'll need to figure out how to handle these other kinds of content. One way we could do it is to simply just store the llm provider's native format without modification (I think that's what currently happens in this PR?). If we do that, then there are two challenges:

- Any Python code that operates on content like images will need to know how to handle the various formats for different LLMs. For example, you could imagine that the app would save an image that's generated by the LLM, or for an image that's uploaded by the user, the app would resize it before sending it to the LLM. A function like this would need to know how to handle the different kinds of representations for each LLM provider.
- The client-side UI code will need to know how to render the different kinds of data sent from the server.

On the other hand, if we were to (internally) normalize the data in Python to a single type of representation, then when `get_messages()` is called, we would need to have a way to transform the data to the right format for the particular LLM provider. That might be another parameter to this method, and/or an option when instantiating the class. Some thoughts here:
- It simplifies the client-side code so that it would only need to handle a single format.
- Coming up with an internal representation that is rich enough to encompass all the LLM providers will take some care.
 

I don't think we necessarily have to have every detail of this figured out right away, but we should design it with the goal of not painting ourselves into a corner.

Reply:
> store the llm provider's native format without modification (I think that's what currently happens in this PR?)

`normalize_message()`/`normalize_message_chunk()` is currently applied before storage, meaning we essentially go from the provider's format to `ChatMessage` for storage. In addition to the benefits you mention, this way we can guarantee a consistent type across user and assistant messages. Also, perhaps more importantly, if we were to instead return the provider format verbatim, you wouldn't get able to pass `.get_messages()` directly into to the `.create()` method.

If and when we do go to add multi-modal input/output, `ChatMessage`'s `content` key will have to change to support a list of content types. I don't think we necessarily have to make a breaking change for this, but we might need to introduce a new class/option that you opt-into which assumes `list[ContentType]` instead of `str`. Also, when it comes to defining `ContentType`, I'm thinking we'll probably roughly follow [OpenAI's convention](https://community.openai.com/t/how-to-get-an-image-described/607957/3)

Reply:
Added a follow up task item about this (the return type of `.get_messages()`)